### ISSUE TITLE
AWS S3 file download logging functionality

### ISSUE THREAD

**Issue by micheleangelofabbri** [dandi/dandi-archive#1424]
There is a potential that International Brain Lab data in NWB format will soon be housed on dandi. A nice to have from the IBL perspective is to know how many times a particular dataset is downloaded. We are currently doing as much with our public data and would like to offer any assistance you may require setting this up on your end with the hopes that this information will one day also be shared with us. Please let me know if there is anything I can do to help.

Link for AWS S3 documentation on how to set up the logs - https://docs.aws.amazon.com/AmazonS3/latest/userguide/ServerLogs.html

**Comment by satra** [comment]
thanks @micheleangelofabbri - i think this is a feature that's been on the roadmap for a while and it would be great to add to the landing page, and perhaps in the listing records as well. i'll leave this to @waxlamp to plan an incorporation on the server side. but the logging should be relatively easy to setup. also given the structure of dandi the atomicity would be at the asset level, so dataset level stats would simply be an aggregate on top of that.

**Comment by satra** [comment]
@waxlamp - just a ping here as this also came up in a conversation with aws that they would like to see some of the access logs. and it does look like a relatively easy setup.

**Comment by waxlamp** [comment]
Got it - prioritizing this.

**Comment by waxlamp** [comment]
@satra: we've come up with a preliminary plan to move on this:

We'll look into turning on the S3 bucket access logs as a first, easy step, and see how that works.

Then, we'll need to understand a bit more deeply exactly how download counts would work. This will be a combination of assessing how we actually tie S3 object downloads back to the Assets (and therefore Versions and Dandisets) they belong to and how to therefore count/report downloads in those entities, and gathering requirements from you (and other stakeholders) on what functionality you actually want to see in the archive.

**Comment by satra** [comment]
thanks @waxlamp - since we use a content-based store the reverse mapping may be difficult unless the log stores where the redirect request came from.

**Comment by danlamanna** [comment]
I think the trickiest part here is tracking zarr downloads, because we only track the top level archive and the requests are essentially range requests. There are 2 approaches that come to mind given that it's prohibitively expensive to store rows related to every single zarr subfile:

1. Consider a zarr download count by bandwidth cost
    We could use some form of S3 object tagging to tag objects with their zarr-id and query usage statistics on some interval. Then the download count is approximate and is computed by `bandwidth / file_size`.
2. Consider a zarr download count when it's "touched"
    We could consider any request to a zarr to be a "download", and again have an approximate download count.

**Comment by yarikoptic** [comment]
I would not over-sophisticate with 1. Pretty much we want 2. as a "zarr download count" may as unique per IP max count of access to any portion within zarr, quantized for each log we consider (are they rotated per our or per day)?  If you give me a sample log, I can craft "oneliner" to express it.

And then we can indeed count a total bandwidth consumed for the archive, or even go for "bandwidth consumed per dandiset" if we dereference assets back to dandisets (insofar it is possible).

**Comment by waxlamp** [comment]
@yarikoptic, your proposal also sounds "over-sophisticated" 😄

At this point, I'd like to focus back on requirements for download counts. I think for non-Zarr assets, it's pretty straightforward: we want to count a download every time the object is requested for download from S3. There are some open questions about how we report that number in the archive, but think a reasonable idea would be to show that download count in the file browser of every Dandiset.

For Zarrs, we can do the same thing but count a download every time any chunk of the Zarr is downloaded (i.e., Dan's Option 2 above). This would reflect the intended real-world usage of Zarr archives (i.e., you're supposed to only download bits and pieces, or bring your computation to the Zarr).

The ideas about uniqueness per IP and total bandwidth seem to be an order of magnitude more complicated than this simple scheme and are perhaps better suited to a separate issue report with its own discussion of usage and requirements. If we can simplify in that way, it will make it much easier for Dan to move this particular issue toward resolution. In particular, it avoids/defers questions of complex data modeling and database usage that would otherwise be required.

**Comment by AlmightyYakob** [comment]
Another way to bridge this gap might be to use the term "accessed" instead of "downloaded" for zarrs, since that's likely what's happening.

**Comment by AlmightyYakob** [comment]
misclick, my bad.

**Comment by yarikoptic** [comment]
> For Zarrs, we can do the same thing but count a download every time any chunk of the Zarr 

it is like counting any TCP packet for http transmission of a large file: would be more reflective of bandwidth and little to have with the number of assets downloaded.

**Comment by AlmightyYakob** [comment]
I'd like to pull back a bit. The point of this issue is for tracking how many times a _dataset is downloaded_.  Even for non-zarr assets, we're talking about _approximating_ the download of a dataset via constituent file downloads, since we really don't have any other meaningful way to track that statistic.

Since downloading isn't really an action that applies to zarrs in this context (they are usually accessed remotely), I think the simplest and most actionable way forward would be to focus on non-zarr assets. If we provide this info to the GUI, we could annotate the download count with a message stating that zarr files aren't included in the download count. If at some point we also want to include zarrs, I think we should provide a "number of times accessed" metric, instead of trying to pigeonhole zarr file access into the "downloads" metric.

**Comment by satra** [comment]
i agree - let's get low hanging fruits done first.  re: zarr - in the future even nwb may use zarr, so we should eventually figure out how to. btw, the aws folks can help us with this as well. they are as much interested in this as we are.

**Comment by waxlamp** [comment]
> i agree - let's get low hanging fruits done first. re: zarr - in the future even nwb may use zarr, so we should eventually figure out how to. btw, the aws folks can help us with this as well. they are as much interested in this as we are.

💯 

> Even for non-zarr assets, we're talking about _approximating_ the download of a dataset via constituent file downloads, since we really don't have any other meaningful way to track that statistic.

I think internally I had already moved on to counting asset downloads, since we don't have (and I can't imagine) a simple definition for a dandiset download count. And furthermore, I think the right thing to do _is_ to count asset downloads, while letting each dandiset report this count in the file browser. That is, I don't see a need to approximate the "dandiset download count" at all (assets are really what count, in some sense, especially under our model of de-duplicating identical asset blobs between dandisets and dandiset versions).

So let's definitely capture the low-hanging fruit, but I'd also like to figure out our true requirements here (as I mentioned above). I think that in turn should involve a discussion of what we want to count, and why. Figuring that out (i.e., putting that to paper so we can all agree on the fundamentals) will help us figure out what to do with Zarr assets.

**Comment by k1o0** [comment]
Coming back to this; the new usage dashboard you have created is fantastic. We've prioritized measuring the impact of our data on the community and would like to track these sorts of metrics over time. In particular, we'd like to measure the impact of publications, hackathons, and conference proceedings on dataset access. Would it be possible to access these stats via an API, or perhaps just to add link to a CSV download?

**Comment by yarikoptic** [comment]
@k1o0 thanks for the feedback!  All the data used by the site frontend AFAIK is coming from the https://github.com/dandi/access-summaries/ repository, so clone or access otherwise (github APIs) at your will!  Having said that, exposing access to those tables would indeed be beneficial, hence filed

- https://github.com/dandi/usage-page/issues/224

Meanwhile, overall, I think we can consider this original long standing issue addressed, thus I will close it.  https://github.com/dandi/usage-page/ repo and its issue tracker and PRs could be used for targeted requests.
